vendors/st/boards/stm32l475_discovery/ports/common_io/iot_i2c.c

```diff
@@ -0,0 +1,744 @@
+        case HAL_OK:
+            return IOT_I2C_SUCCESS;
+
+        case HAL_ERROR:
+            return operation == _I2C_READ_OP ? IOT_I2C_READ_FAILED : IOT_I2C_WRITE_FAILED;
+
+        case HAL_BUSY:
+            return IOT_I2C_BUSY;
+
+        case HAL_TIMEOUT:
+            return IOT_I2C_BUS_TIMEOUT;
+    }
+
+    /* TODO: it is better to assert(0) here. */
+    return IOT_I2C_SUCCESS;
+}
+
+static int32_t _doMasterTransfer( IotI2CDescriptor_t * pI2cDescriptor,
+                                  uint8_t * const pucBuffer,
+                                  size_t xBytes,
+                                  HAL_StatusTypeDef ( * pMasterSync )( I2C_HandleTypeDef * hi2c,
+                                                                       uint16_t DevAddress,
+                                                                       uint8_t * pData,
+                                                                       uint16_t Size,
+                                                                       uint32_t Timeout ),
+                                  HAL_StatusTypeDef ( * pMasterIT )( I2C_HandleTypeDef * hi2c,
+                                                                     uint16_t DevAddress,
+                                                                     uint8_t * pData,
+                                                                     uint16_t Size ),
+                                  HAL_StatusTypeDef ( * pMasterSequentialIT )( I2C_HandleTypeDef * hi2c,
+                                                                               uint16_t DevAddress,
+                                                                               uint8_t * pData,
+                                                                               uint16_t Size,
+                                                                               uint32_t XferOptions ) )
+{
+    HAL_StatusTypeDef halStatus = HAL_OK;
+    uint8_t ucAsyncCall = 1;
+
+    uint8_t ongoingTrans = ( pI2cDescriptor->xHandle.PreviousState & _I2C_STATE_BUSY_MASK ) > 0;
+
+    /* Do not send stop condition. */
+    if( _isFlagSet( pI2cDescriptor, _FLAG_NO_STOP ) )
+    {
+        /* If the previous state is ready, then there is no ongoing transaction; otherwise, there is. */
+        halStatus = pMasterSequentialIT( &( pI2cDescriptor->xHandle ), pI2cDescriptor->usSlaveAddr, pucBuffer, xBytes, ongoingTrans ? I2C_NEXT_FRAME : I2C_FIRST_FRAME );
+
+        /* Clear no stop flag. */
+        pI2cDescriptor->flags &= ~_FLAG_NO_STOP;
+    }
+    else
+    {
+        if( ongoingTrans )
+        {
+            /* There is an ongoing transaction; end the transaction. */
+            halStatus = pMasterSequentialIT( &( pI2cDescriptor->xHandle ), pI2cDescriptor->usSlaveAddr, pucBuffer, xBytes, I2C_LAST_FRAME );
+        }
+        else
+        {
+            if( pMasterSync != NULL )
+            {
+                /* No ongoing transaction; simple make a blocking call. */
```

Comment:
"simply"

Reply:
Thank you for reviewing this so carefully!